### ISSUE TITLE
GHC-9.8 backwards compatibility?

### ISSUE THREAD

**Issue by gergoerdi** [ucsd-progsys/liquidhaskell#2379]
I looked at what it would take to still support GHC 9.8 with `develop` as of 5595507d943cb00f92674c19a5cbafcd93dff92b and it's so little that I think it would be feasible with just a little CPP usage to keep supporting 9.8. For comparison, Clash, which is a similarly involved user of the GHC API, [manages to support all GHC versions between 8.10 and 9.10](https://github.com/clash-lang/clash-compiler/tree/5706eafca799ae04fda0ee7d666a40b6c0e7f22b?tab=readme-ov-file#ghc-compatibility). 

- [x] `foldl'` is not exported by the `Prelude`, so has to be imported from `Data.List`. A triviality that affects 8 files. The imports can be guarded by CPP macros if we want to avoid "redundant imports" warnings on GHC 9.10.

- [x] `Unique`s are built from `Int`s, not `Word64`s. Handling this difference can be centralized by adding a type alias and some utility functions to some new  `Liquid.GHC.API.Compat` module.

- [x]  Instead of `GHC.Internal.Foldable` and `GHC.Internal.Real`, we have `Data.Foldable` and `GHC.Real`.  Again, can be worked around by defining aliases in `Liquid.GHC.API.Compat`.

- [x] Differences in AST annotations / TTG expansion points: This is very annoying, because it requires CPP branching at five places across three modules. 

- [x] `tcValBinds` has a slightly different type, requiring a change in `Language.Haskell.Liquid.GHC.Misc.withWiredIn`. 

Overall, I don't think it would be that big a burden to keep GHC 9.8 around.

**Comment by gergoerdi** [comment]
I've pushed the change for the first three issues to https://github.com/gergoerdi/liquidhaskell/tree/ghc-9.8-compat

I've got code locally that "fixes" the remaining problems (by breaking the GHC 9.10 compatibility), so at least we can know there's no other "unknown unknowns" lurking.

**Comment by facundominguez** [comment]
On the side of costs, we would need to make sure patches pass tests in both compiler versions. The hardest thing to support is probably going to be the move of many definitions from `base` to `ghc-internals`, which affects the annotations in the `liquidhaskell` package and wired-in names in the code.

I'd prefer having a separate branch for ghc-9.8 and backporting fixes there when needed, rather than making harder to accept contributions to the `develop` branch. It seems to me like it would allow to better share the maintenance effort.

For the reference, here is the issue (#2307) and the PR (#2316) about the upgrade to 9.10.1.

**Comment by gergoerdi** [comment]
> I'd prefer having a separate branch for ghc-9.8 and backporting fixes there when needed

So would you be happy with me starting a new branch off current `develop`, where I'd push all the GHC 9.8 compatibility changes (*without* CPP-guarding them) as a new baseline for this branch going forward? Otherwise, I'm worried things will keep drifting further and further apart.

**Comment by facundominguez** [comment]
> where I'd push all the GHC 9.8 compatibility changes (without CPP-guarding them) as a new baseline for this branch going forward?

That works. There are two strategies to maintain that:
1. Add backports of patches from develop on top.
2. Rebase the compatibility changes every now and then on top of the latest `develop`.

In either case, I suppose CI will need to be reverted to use GHC 9.8 on that branch.

One other thought that comes to mind: it might be worth encapsulating the compatibility changes so the differences between `develop` and `develop-ghc-9.8` spread over as little files as possible.

**Comment by gergoerdi** [comment]
> One other thought that comes to mind: it might be worth encapsulating the compatibility changes so the differences between `develop` and `develop-ghc-9.8` spread over as little files as possible.

#2380 is a good shot at this. I could remove just the CPP parts and then the only diff on the GHC-9.8 branch will be in in what are currently CPP-guarded in #2380. Does this make sense?

**Comment by facundominguez** [comment]
> Does this make sense?

Yes, I think #2380 works. Removing the CPP guards also sounds good to me.

I expect the encapsulation to be harder for dealing with annotations, as much of those in `liquidhaskell` would need to be rewritten, unless we can get rid of the need to fully qualify assumption names.

**Comment by gergoerdi** [comment]
Could you make a new `ghc-9.8-backport` branch so I could open a PR to it?

**Comment by facundominguez** [comment]
:+1: there is a branch `develop-ghc-9.8` now.

**Comment by gergoerdi** [comment]
- [ ] GHC 9.8 has no `ghc-internals`, so the annotations in `liquidhaskell` package point to the wrong module names

**Comment by facundominguez** [comment]
#2430 is removing the explicit dependency on `ghc-internals`. Perhaps that makes support for GHC 9.8 easier.

**Comment by facundominguez** [comment]
Should we close this task as completed?

**Comment by facundominguez** [comment]
hm, probably we should document the existence of `develop-ghc-9.8` somewhere.
